In this project, we use Northwind database, which was orinally created by Microsofts and contains the sales data for a fictitious company called “Northwind Traders,” which imports and exports specialty foods from around the world. As a team of Data Analyst team , we use SQL to answer several business questions: <br>
**Task 1**: Track the sale situation of each quarter of year, each product, each country <br>
**Task 2**: Find the number of loyal customers and performance of employees on the first quarter of 1998, based on the policy of company <br>
**Task 3**: Find products which is often ordered together by customers <br>

The Northwind dataset includes sample data for the following. <br>
**Suppliers**: Suppliers and vendors of Northwind <br>
**Customers**: Customers who buy products from Northwind <br>
**Employees**: Employee details of Northwind traders <br>
**Products**: Product information <br>
**Shippers**: The details of the shippers who ship the products from the traders to the end-customers <br>
**Orders** and **OrderDetails**: Sales Order transactions taking place between the customers & the company <br>

In [1]:
import pyodbc
import sqlalchemy
import matplotlib

In [2]:
%load_ext sql

In [3]:
%sql mssql+pyodbc://sa:12342345@sql

First of all, we check tables and columns in this database. 

In [17]:
%%sql
SELECT *
FROM Northwind.information_schema.tables
WHERE TABLE_TYPE='BASE TABLE';

 * mssql+pyodbc://sa:***@sql
Done.


[('Northwind', 'dbo', 'Employees', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Categories', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Shippers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Suppliers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Orders', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Products', 'BASE TABLE'),
 ('Northwind', 'dbo', 'OrderDetails', 'BASE TABLE'),
 ('Northwind', 'dbo', 'CustomerCustomerDemo', 'BASE TABLE'),
 ('Northwind', 'dbo', 'CustomerDemographics', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Region', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Territories', 'BASE TABLE'),
 ('Northwind', 'dbo', 'EmployeeTerritories', 'BASE TABLE'),
 ('Northwind', 'dbo', 'sysdiagrams', 'BASE TABLE')]

In [5]:
%%sql
select *
from Northwind.information_schema.columns c join Northwind.information_schema.tables t on c.TABLE_NAME= t.TABLE_NAME
where TABLE_TYPE = 'BASE TABLE';

 * mssql+pyodbc://sa:***@sql
Done.


[('Northwind', 'dbo', 'Customers', 'CustomerID', 1, None, 'NO', 'nchar', 5, 10, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'CompanyName', 2, None, 'NO', 'nvarchar', 40, 80, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'ContactName', 3, None, 'YES', 'nvarchar', 30, 60, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'ContactTitle', 4, None, 'YES', 'nvarchar', 30, 60, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'Address', 5, None, 'YES', 'nvarchar', 60, 120, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'City', 6, None, 'YES', 'nvarchar', 15, 30, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'Region', 7, None, 'YES', 'nvarchar', 15, 30, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'PostalCode', 8, None, 'YES', 'nvarchar', 10, 20, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'Country', 9, None, 'YES', 'nvarchar', 15, 30, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'Phone', 10, None, 'YES', 'nvarchar', 24, 48, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Customers', 'Fax', 11, None, 'YES', 'nvarchar', 24, 48, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Customers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Suppliers', 'SupplierID', 1, None, 'NO', 'int', None, None, 10, 10, 0, None, None, None, None, None, None, None, None, None, None, 'Northwind', 'dbo', 'Suppliers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Suppliers', 'CompanyName', 2, None, 'NO', 'nvarchar', 40, 80, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Suppliers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Suppliers', 'ContactName', 3, None, 'YES', 'nvarchar', 30, 60, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Suppliers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Suppliers', 'ContactTitle', 4, None, 'YES', 'nvarchar', 30, 60, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Suppliers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Suppliers', 'Address', 5, None, 'YES', 'nvarchar', 60, 120, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'Northwind', 'dbo', 'Suppliers', 'BASE TABLE'),
 ('Northwind', 'dbo', 'Suppliers', 'City', 6, None, 'YES', 'nvarchar', 15, 30, None, None, None, None, None, None, 'UNICODE', None, None, 'SQL_Latin1_General_CP1_CI_AS', None, None, None, 'North

**Task 1**: Track the sale situation of each quarter of year, each product (show top 10), each country

In [6]:
%%sql
select min(RequiredDate) as start_date, max(RequiredDate) as end_date 
from Northwind.dbo.Orders;

 * mssql+pyodbc://sa:***@sql
Done.


[(datetime.datetime(1996, 7, 24, 0, 0), datetime.datetime(1998, 6, 11, 0, 0))]

In [7]:
%%sql
WITH sum_per_quarter AS (
    select year(o.RequiredDate) as year, DATEPART(QUARTER, o.RequiredDate) as quarter, cast(sum(od.UnitPrice*od.Quantity*(1-od.Discount)) as decimal (10,2)) as sum_turnover
    from Northwind.dbo.OrderDetails od join Northwind.dbo.Orders o on od.OrderID=o.OrderID
    group by year(o.RequiredDate), DATEPART(QUARTER, o.RequiredDate)
)
select year, quarter, sum_turnover, cast((sum_turnover/(LAG(sum_turnover) OVER (ORDER BY year, quarter)) -1)*100 as decimal(5,2)) as percent_of_growth
from sum_per_quarter
ORDER BY year, quarter;

 * mssql+pyodbc://sa:***@sql
Done.


[(1996, 3, Decimal('53468.77'), None),
 (1996, 4, Decimal('111870.53'), Decimal('109.23')),
 (1997, 1, Decimal('139631.18'), Decimal('24.81')),
 (1997, 2, Decimal('150636.26'), Decimal('7.88')),
 (1997, 3, Decimal('134765.37'), Decimal('-10.54')),
 (1997, 4, Decimal('175874.39'), Decimal('30.50')),
 (1998, 1, Decimal('263539.34'), Decimal('49.85')),
 (1998, 2, Decimal('236007.20'), Decimal('-10.45'))]

This database records data form 7/1996 to 6/1998. Generally, sales've increased over the quarters. Noticeably, on the first quarter of 1998, sales increased by 49% compared to the last quarter. It seems that sales decreased on the second quarter, but it is reason is that database didn't contain all orders in this quarter. 

Now, we track sale situation of each product in 1997. First, we need to know how many products the company has.

In [27]:
%%sql 
SELECT Discontinued, count(*) as count
FROM Northwind.dbo.Products
GROUP BY Discontinued;

 * mssql+pyodbc://sa:***@sql
Done.


[(False, 69), (True, 8)]

In [37]:
%%sql
WITH top_product AS (
    SELECT sum(od.UnitPrice*od.Quantity*(1-od.Discount)) as sum_revenue, od.ProductID
    FROM Northwind.dbo.OrderDetails od join Northwind.dbo.Orders o on o.OrderID = od.OrderID
    WHERE year(o.RequiredDate)=1997
    GROUP BY od.ProductID
)
SELECT TOP 20 PERCENT p.ProductID, p.ProductName, p.CategoryID, s.SupplierID, s.CompanyName, cast(tp.sum_revenue as decimal(10,2)) as sum, 
    tp.percent_of_total, sum(tp.percent_of_total) OVER (ORDER BY tp.sum_revenue DESC) as cucum_percent, p.Discontinued
FROM (
    SELECT *, CAST(sum_revenue/(SELECT sum(sum_revenue) FROM top_product)*100 as decimal(10,2)) as percent_of_total
    FROM top_product
) as tp 
JOIN Northwind.dbo.Products p ON tp.ProductID = p.ProductID
JOIN Northwind.dbo.Suppliers s ON s.SupplierID = p.SupplierID
ORDER BY tp.sum_revenue DESC;

 * mssql+pyodbc://sa:***@sql
Done.


[(38, 'Côte de Blaye', 1, 18, 'Aux joyeux ecclésiastiques', Decimal('51569.59'), Decimal('8.58'), Decimal('8.58'), False),
 (59, 'Raclette Courdavault', 4, 28, 'Gai pâturage', Decimal('32769.55'), Decimal('5.45'), Decimal('14.03'), False),
 (56, 'Gnocchi di nonna Alice', 5, 26, 'Pasta Buttini s.r.l.', Decimal('31958.00'), Decimal('5.32'), Decimal('19.35'), False),
 (29, 'Thüringer Rostbratwurst', 6, 12, 'Plutzer Lebensmittelgroßmärkte AG', Decimal('28058.43'), Decimal('4.67'), Decimal('24.02'), True),
 (60, 'Camembert Pierrot', 4, 28, 'Gai pâturage', Decimal('22413.48'), Decimal('3.73'), Decimal('27.75'), False),
 (62, 'Tarte au sucre', 3, 29, "Forêts d'érables", Decimal('20811.33'), Decimal('3.46'), Decimal('31.21'), False),
 (51, 'Manjimup Dried Apples', 7, 24, "G'day, Mate", Decimal('17508.02'), Decimal('2.91'), Decimal('34.12'), False),
 (69, 'Gudbrandsdalsost', 4, 15, 'Norske Meierier', Decimal('14905.80'), Decimal('2.48'), Decimal('36.60'), False),
 (18, 'Carnarvon Tigers', 8, 7, 'Pavlova, Ltd.', Decimal('14525.00'), Decimal('2.42'), Decimal('39.02'), False),
 (17, 'Alice Mutton', 6, 7, 'Pavlova, Ltd.', Decimal('14123.85'), Decimal('2.35'), Decimal('41.37'), True),
 (53, 'Perth Pasties', 6, 24, "G'day, Mate", Decimal('13832.13'), Decimal('2.30'), Decimal('43.67'), True),
 (28, 'Rössle Sauerkraut', 7, 12, 'Plutzer Lebensmittelgroßmärkte AG', Decimal('13691.48'), Decimal('2.28'), Decimal('45.95'), True),
 (72, 'Mozzarella di Giovanni', 4, 14, 'Formaggi Fortini s.r.l.', Decimal('12427.38'), Decimal('2.07'), Decimal('48.02'), False),
 (26, 'Gumbär Gummibärchen', 3, 11, 'Heli Süßwaren GmbH & Co. KG', Decimal('11225.66'), Decimal('1.87'), Decimal('49.89'), False),
 (43, 'Ipoh Coffee', 1, 20, 'Leka Trading', Decimal('11069.90'), Decimal('1.84'), Decimal('51.73'), False),
 (27, 'Schoggi Schokolade', 3, 11, 'Heli Süßwaren GmbH & Co. KG', Decimal('10974.00'), Decimal('1.83'), Decimal('53.56'), False)]

In [35]:
%%sql
SELECT CategoryName
FROM Northwind.dbo.Categories 
WHERE CategoryID = 6;

 * mssql+pyodbc://sa:***@sql
Done.


[('Meat/Poultry',)]

In this result, we show only top 20 percent of products, which account for more than 50% of revenue in 1997. We notice that there were 3 products in CategoryID 6 (Meat/Poultry) but all of them were discontinued. We should discuss with other departments to find the reason of this discontinuation.

In [12]:
query= '''
select c.Country, cast(sum(od.UnitPrice*od.Quantity*(1-od.Discount)) as decimal (10,2)) as sum_turnover
from Northwind.dbo.OrderDetails od 
join Northwind.dbo.Orders o on od.OrderID = o.OrderID
join Northwind.dbo.Customers c on c.CustomerID=o.CustomerID
where year(o.RequiredDate) = 1997
group by c.Country
order by sum(od.UnitPrice*od.Quantity*(1-od.Discount)) desc;'''

In [13]:
result=%sql $query
result

 * mssql+pyodbc://sa:***@sql
Done.


[('USA', Decimal('118376.01')),
 ('Germany', Decimal('106011.79')),
 ('Austria', Decimal('52665.05')),
 ('Brazil', Decimal('46741.27')),
 ('France', Decimal('38514.02')),
 ('Canada', Decimal('33529.86')),
 ('Sweden', Decimal('29392.79')),
 ('Venezuela', Decimal('25910.97')),
 ('UK', Decimal('25153.46')),
 ('Ireland', Decimal('23134.63')),
 ('Denmark', Decimal('22331.05')),
 ('Switzerland', Decimal('18380.82')),
 ('Mexico', Decimal('14349.28')),
 ('Finland', Decimal('12461.41')),
 ('Belgium', Decimal('11406.48')),
 ('Portugal', Decimal('7191.24')),
 ('Italy', Decimal('6440.01')),
 ('Spain', Decimal('4087.55')),
 ('Argentina', Decimal('1804.10')),
 ('Norway', Decimal('1758.40')),
 ('Poland', Decimal('1267.00'))]

**Task 2**: Find the number of loyal customers and performance of employees on the first quarter of 1998, based on the policy of company <br>
*Scenario*: Since 1998, Northwind has a policy called "Loyal customers", which is applied for customers in USA, UK and France. <br>
- If customers bought total value 10000 or above in 1997, they will become "Gold customer" and get discount 5% for all orders in 1998. <br>
- If customers bought total value 10000 or above in 1997, they will become "Silver customer" and get discount 2% for all orders in 1998. <br>
Based on this policy, we find number of gold customers, silver customers and find top 3 employees has best performance in the first quarter of 1998. 

In [39]:
%%sql
WITH customer_type (CustomerID, Value1997,Level1998) AS (
    SELECT o.CustomerID as CustomerID, sum(od.UnitPrice*od.Quantity*(1-od.Discount)) as Value1997, (
        CASE WHEN sum(od.UnitPrice*od.Quantity*(1-od.Discount)) >= 10000 THEN '2. Gold customer'
            WHEN sum(od.UnitPrice*od.Quantity*(1-od.Discount)) >=5000 THEN '1. Silver customer'
            ELSE '0. Normal customer'
        END
    ) as Level1998
    FROM Northwind.dbo.OrderDetails od
    JOIN Northwind.dbo.Orders o ON o.OrderID=od.OrderID 
    JOIN Northwind.dbo.Customers c ON c.CustomerID = o.CustomerID
    WHERE (c.Country='USA' or c.Country= 'UK' or c.Country='France') AND YEAR(o.RequiredDate)=1997
    GROUP BY o.CustomerID
)

SELECT Level1998 AS type_of_customer, count(*) AS Number
FROM customer_type
GROUP BY Level1998;


 * mssql+pyodbc://sa:***@sql
Done.


[('0. Normal customer', 17),
 ('1. Silver customer', 9),
 ('2. Gold customer', 3)]

The number of gold customers is 3 and the number of silver customers is 9. These amount is acceptable, therefore, this policy haven't to modify. By the end of 1998, we should evaluate its effectiveness, then decide if this policy should be applied for customers all over the world. 

In [14]:
%%sql
WITH value1997 AS (
    SELECT c.CustomerID as CustomerID, sum(od.UnitPrice*od.Quantity*(1-od.Discount)) as total_of_customer_1997, (
        CASE
        WHEN c.CustomerID IN (SELECT CustomerID FROM Northwind.dbo.Customers WHERE Country IN ('USA', 'UK', 'France')) THEN (
            CASE
            WHEN sum(od.UnitPrice*od.Quantity*(1-od.Discount)) >= 10000 THEN 0.05
            WHEN sum(od.UnitPrice*od.Quantity*(1-od.Discount)) >= 5000 THEN 0.02
            ELSE 0
            END
        )
        ELSE 0
        END
    ) AS customer_discount
    FROM Northwind.dbo.Orders o
    JOIN Northwind.dbo.OrderDetails od ON o.OrderID = od.OrderID 
    JOIN Northwind.dbo.Customers c ON c.CustomerID = o.CustomerID
    WHERE YEAR(o.RequiredDate)= 1997
    GROUP BY c.CustomerID
),
order_total AS (
    SELECT o.OrderID as OrderID, sum(od.UnitPrice*od.Quantity*(1-od.Discount)) as total_of_order
    FROM Northwind.dbo.Orders o JOIN Northwind.dbo.OrderDetails od ON o.OrderID = od.OrderID
    WHERE o.RequiredDate >= '1998-01-01' AND o.RequiredDate < '1998-04-01'
    GROUP BY o.OrderID
),
order_total_after_discount AS (
    SELECT o.OrderID as OrderID, o.EmployeeID as EmployeeID, (
        CASE WHEN o.OrderDate >='1998-01-01' AND o.CustomerID IN (SELECT CustomerID FROM value1997) THEN ot.total_of_order*(1-v.customer_discount) 
        ELSE ot.total_of_order
        END
    ) as after_discount
    FROM Northwind.dbo.Orders o 
    JOIN order_total ot ON o.OrderID= ot.OrderID
    LEFT JOIN value1997 v ON o.CustomerID = v.CustomerID
),
employee_total AS (
    SELECT EmployeeID, sum(after_discount) as total_of_turnover
    FROM order_total_after_discount
    GROUP BY EmployeeID
)
SELECT TOP (3) e.EmployeeID, e.FirstName, e.LastName, e.Title, 
    cast(et.total_of_turnover as decimal(10,2)) as turnover_of_employee, 
    cast((100*et.total_of_turnover/(SELECT sum(total_of_turnover) FROM employee_total)) as decimal(10,2)) as percent_of_quarter
FROM Northwind.dbo.Employees e JOIN employee_total et ON e.EmployeeID = et.EmployeeID
ORDER BY et.total_of_turnover DESC;


 * mssql+pyodbc://sa:***@sql
Done.


[(3, 'Janet', 'Leverling', 'Sales Representative', Decimal('61200.59'), Decimal('23.40')),
 (4, 'Margaret', 'Peacock', 'Sales Representative', Decimal('43346.45'), Decimal('16.57')),
 (2, 'Andrew', 'Fuller', 'Vice President, Sales', Decimal('35391.48'), Decimal('13.53'))]

On the first quarter of 1998, top 3 sale employees includes: Janet Leverling (23,4% of total revenue), Margaret Peacock (16,57% of total revenue) and Andrew Fuller - the Vice President. 

**Task 3**: Find products which is often ordered together by customers <br>
Should company have discount when customers buy a "combo", which contains certain product pairs? To answer this question, we need to find which products are often ordered together within a order.

In [16]:
%%sql
WITH pairs AS (
    SELECT od1.ProductID AS ProductID_1, od2.ProductID AS ProductID_2, od1.OrderID AS OrderID 
    FROM Northwind.dbo.OrderDetails od1 
    JOIN Northwind.dbo.OrderDetails od2 ON od1.OrderID = od2.OrderID
    WHERE od1.ProductID < od2.ProductID
)
SELECT TOP (3) cp.ProductID_1, p1.ProductName as ProductName_1, p1.CategoryID as Category_1, cp.ProductID_2, p2.ProductName as ProductName_2, p2.CategoryID as Category_2, cp.number_of_order
FROM (
    SELECT count(*) AS number_of_order, ProductID_1, ProductID_2
    FROM pairs 
    GROUP BY ProductID_1, ProductID_2
) as cp
LEFT JOIN Northwind.dbo.Products p1 ON cp.ProductID_1 = p1.ProductID
LEFT JOIN Northwind.dbo.Products p2 ON cp.ProductID_2 = p2.ProductID
ORDER BY cp.number_of_order DESC;

 * mssql+pyodbc://sa:***@sql
Done.


[(21, "Sir Rodney's Scones", 3, 61, "Sirop d'érable", 2, 8),
 (16, 'Pavlova', 3, 31, 'Gorgonzola Telino', 4, 7),
 (16, 'Pavlova', 3, 60, 'Camembert Pierrot', 4, 6)]

In [15]:
%%sql
SELECT count(*)
FROM Northwind.dbo.Orders;

 * mssql+pyodbc://sa:***@sql
Done.


[(830,)]

The above result show the most common pairs of products. However, the greatest number of order is 8 and very small compared to number of total order (830), then we don't recommend to apply combo discount. 